Implement curation tasks for Configurable Workflow

mwoodiupui opened the pull request:
## References
* Fixes #5424 (DS-2058)
* May affect #3144.
* Possibly affected by #3028.

## Description
Port DS-2058 to 7x:  implement attachment of Curation Tasks to Configurable Workflow steps.

## Instructions for Reviewers
List of changes in this PR:
* New classes to represent the task mapping (configured from the existing XML document) and to execute tasks as submissions progress through workflow.  Adapted from the old fixed-workflow code (which is to be removed in #3144).
* New XML Schema document used to generate a validating parser and to validate the configuration.
* Mojohaus' jaxb2-maven-plugin used to generate the parser.
* Small change to the configuration document:  the empty taskset "none" is now explicitly defined, to make the document schema-valid.

**Include guidance for how to test or review your PR.** This may include: steps to reproduce a bug, screenshots or description of a new feature, or reasons behind specific changes. 

## Checklist
- [ ] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

mwoodiupui committed:
Clean up some commentary

mwoodiupui committed:
[DS-2058] First cut at adding curation to Configurable Workflow

mwoodiupui committed:
[DS-2058] Replace unwarranted assumptions about XML attribute order with named attributes

tdonohue committed:
Rename to RegistryUpdater so that it comes alphabetically *after* GroupServiceInitializer as Flyway runs callbacks alphabetically

mwoodiupui committed:
[DS-2058] Port to 7x.  Formalize parser.  Add tests.

mwoodiupui committed:
Merge branch 'main' into DS-2058-7x

mwoodiupui committed:
[DS-2058] Satisfy checkstyle.

mwoodiupui committed:
[DS-2058] Enable autowiring and point to the configuration document.

mwoodiupui committed:
[DS-2058] Make <mapping taskset='none'/> valid by defining empty dummy taskset.

mwoodiupui committed:
Add much missing javadoc.  Elaborate or clarify quite a bit too. [DS-2058]

mwoodiupui committed:
[DS-2058] Supply more missing Javadoc

mwoodiupui committed:
Merge remote-tracking branch 'upstream/main' into DS-2058-7x

mwoodiupui committed:
[DS-2058] Correct path to workflow curation configuration.

Also clean up an annoying typo in a comment.

mwoodiupui committed:
[DS-2058] Catch up with removal of traditional workflow in main.

mwoodiupui committed:
[DS-2058] findTaskSet should *always* return a TaskSet.

mwoodiupui committed:
[DS-2058] Code cleanup, documentation.

mwoodiupui committed:
[DS-2058] Failing attempt at an integration test.

mwoodiupui committed:
[DS-2058] Scrap integration test, make it a unit test, fake claimed tasks.

mwoodiupui committed:
[DS-2058] Remove impossible tests of Curator state -- Curator is reset by service.

mwoodiupui committed:
[DS-2058] Satisfy Checkstyle

mwoodiupui commented:
Rather large, but some of it is missing or inadequate Javadoc that I filled in while trying to understand the environment in which this runs.

mwoodiupui committed:
Merge remote-tracking branch 'upstream/main' into DS-2058-7x

tdonohue commented on dspace-api/src/main/java/org/dspace/workflow/FlowStep.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * A workflow step.
+ */
+public class FlowStep {
```
I'd appreciate it if we could improve the JavaDocs here.  This currently implies that a FlowStep is how you define a new Workflow Step.  But, that would then make it equivalent to `org.dspace.xmlworkflow.state.Step`  & we know that's not accurate.

It seems to me that this class is not actually defining a Workflow Step.  It's instead defining a mapping (is that the right word?) of Curation Task(s) to a Workflow Step defined in `workflow.xml`.  At least that's what I *think* is happening.

My point is that the lack of explanation here makes this very confusing.  I think the role of this FlowStep class is to map to a `<flowStep>` in `workflow-curation.xml`....but, it's unclear how that overlaps with the configurable workflow in `workflow.xml`...and whether the steps must share names or not.

mwoodiupui replied:
This reflects my own confusion when I began updating these classes.  You are correct:  these objects embody the relationships between workflow steps and curation tasks.  I hope that the commentary is clearer now.

tdonohue commented on dspace-api/src/main/java/org/dspace/workflow/Task.java:
```diff
@@ -0,0 +1,65 @@
+/**
+ * A workflow task.
+ */
+public class Task {
```
Same here, improve JavaDocs please.  This isn't a Workflow "task" in the same manner as a ClaimedTask or PooledTask.  Instead, this seems to be defining a *Curation Task* that is mapped to a Workflow step.

tdonohue commented on dspace-api/src/main/java/org/dspace/workflow/TaskSet.java:
```diff
@@ -0,0 +1,30 @@
+/**
+ * A collection of workflow tasks.
+ */
+public class TaskSet {
```
Same here, improve javadocs please.  Again, the phrase "workflow tasks" is probably incorrect.  These are Workflow *Curation* Tasks.

mwoodiupui replied:
Right.  I've tried to make explicit which bits are Curation objects and which are Workflow objects.

tdonohue commented on dspace/config/workflow-curation.xml:
```diff
@@ -61,6 +61,9 @@
        </task>
      </flowstep>
    </taskset>
```
Based on the tests in this PR, we might want to improve the comments/docs for the `workflow-curation.xml`.  The comment at the beginning of this config file still references "DSpace provides three workflow steps" and tells you how to configure `workflow-curation.xml` based on those three steps.  It looks like that behavior has changed now.  Now, you need to know the Step "name" in `workflow.xml` and match it up in `workflow-curation.xml`... so, we should better document that in this config file.

mwoodiupui replied:
Argh, another place that I meant to revisit.  The documentation in the `.xsd` was closer but still referred to the old 3-step workflow model (since the text was initially copied from `workflow-curation.xml`).  Both should be more accurate and less murky now.

tdonohue reviewed (changes requested):
@mwoodiupui : I gave this a quick code review today.  Overall, the code looks good, but I think we could improve some of the JavaDocs & the comments in `workflow-curation.xml` to describe how this works with Configurable Workflow. See inline comments below.

I haven't had a chance to test this out yet, but will do so hopefully early next week.

tdonohue reviewed (changes requested):
@mwoodiupui : I built this today to try testing it, but when I start up Tomcat I get Spring errors relating to the `NoAction` class... It's a very long exception, but here's the last three `Caused By` clauses which seem to summarize the issues:
```
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'noAction' defined in URL [jar:file:/dspace/webapps/server/WEB-INF/lib/dspace-api-7.0-beta6-SNAPSHOT.jar!/spring/spring-dspace-addon-workflow-services.xml]: Cannot create inner bean 'org.dspace.xmlworkflow.state.actions.processingaction.NoAction#2668fa28' of type [org.dspace.xmlworkflow.state.actions.processingaction.NoAction] while setting bean property 'processingAction'; nested exception is org.springframework.beans.factory.CannotLoadBeanClassException: Cannot find class [org.dspace.xmlworkflow.state.actions.processingaction.NoAction] for bean with name 'org.dspace.xmlworkflow.state.actions.processingaction.NoAction#2668fa28' defined in URL [jar:file:/dspace/webapps/server/WEB-INF/lib/dspace-api-7.0-beta6-SNAPSHOT.jar!/spring/spring-dspace-addon-workflow-services.xml]; nested exception is java.lang.ClassNotFoundException: org.dspace.xmlworkflow.state.actions.processingaction.NoAction
                at org.springframework.beans.factory.support.BeanDefinitionValueResolver.resolveInnerBean(BeanDefinitionValueResolver.java:389)
                at org.springframework.beans.factory.support.BeanDefinitionValueResolver.resolveValueIfNecessary(BeanDefinitionValueResolver.java:127)
                at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyPropertyValues(AbstractAutowireCapableBeanFactory.java:1699)
                at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1444)
                at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:594)
                at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:517)
                at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:341)
                at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
                at org.springframework.beans.factory.support.BeanDefinitionValueResolver.resolveReference(BeanDefinitionValueResolver.java:330)
                ... 42 more
        Caused by: org.springframework.beans.factory.CannotLoadBeanClassException: Cannot find class [org.dspace.xmlworkflow.state.actions.processingaction.NoAction] for bean with name 'org.dspace.xmlworkflow.state.actions.processingaction.NoAction#2668fa28' defined in URL [jar:file:/dspace/webapps/server/WEB-INF/lib/dspace-api-7.0-beta6-SNAPSHOT.jar!/spring/spring-dspace-addon-workflow-services.xml]; nested exception is java.lang.ClassNotFoundException: org.dspace.xmlworkflow.state.actions.processingaction.NoAction
                at org.springframework.beans.factory.support.AbstractBeanFactory.resolveBeanClass(AbstractBeanFactory.java:1477)
                at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:489)
                at org.springframework.beans.factory.support.BeanDefinitionValueResolver.resolveInnerBean(BeanDefinitionValueResolver.java:374)
                ... 50 more
        Caused by: java.lang.ClassNotFoundException: org.dspace.xmlworkflow.state.actions.processingaction.NoAction
                at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1372)
                at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1195)
                at java.base/java.lang.Class.forName0(Native Method)
                at java.base/java.lang.Class.forName(Class.java:398)
                at org.springframework.util.ClassUtils.forName(ClassUtils.java:285)
                at org.springframework.beans.factory.support.AbstractBeanDefinition.resolveBeanClass(AbstractBeanDefinition.java:456)
                at org.springframework.beans.factory.support.AbstractBeanFactory.doResolveBeanClass(AbstractBeanFactory.java:1542)
                at org.springframework.beans.factory.support.AbstractBeanFactory.resolveBeanClass(AbstractBeanFactory.java:1469)
                ... 52 more
```

Obviously, this class does exist, so, I'm not sure I understand exactly what is going on here.  I haven't had time to dig in deeply, so I just wanted to be sure to report it. (So that you can see if it's repeatable on your end, or not.)

mwoodiupui committed:
[DS-2058] Clarify doc commentary.

mwoodiupui committed:
[DS-2058] Add missing license to new package-info file.

abollini commented on dspace-api/src/main/java/org/dspace/core/AbstractHibernateDAO.java:
```diff
@@ -209,7 +209,14 @@ public Query createQuery(Context context, String query) throws SQLException {
      * so only use when the criteria state uniqueness in the database.
      * @param criteriaQuery JPA criteria
-     * @return a DAO specified by the criteria
+     * @param cacheable whether or not this query should be cacheable.
```
the context param is missing in the javadoc

mwoodiupui replied:
Fixed.

abollini reviewed (commented):
I guess to have find the issue hit by @tdonohue it is caused by the curation workflow test spring configuration to be activated by default instead than only during test.

Except for that the code looks generally good to me, I would just prefer to replace the old XmlWorkflowServiceImplTest with an Integration test done on the REST layer. You can configure a custom workflow for a special collection (handle 123456789/collection-workflow-curation) that should trigger the noaction curation.
The TaskRestRepositoriesIT contains tests that move workflow over the different step this should be the basis to verify that the curation is triggered when a specific step is reach and that the result of the curation influence the workflow.

mwoodiupui committed:
[DS-2058] Move test scaffolding from main to test resources.

mwoodiupui committed:
[DS-2058] Javadoc fixes.

mwoodiupui committed:
[DS-2058] Rename message to match the class that uses it.

Also create a new message and stop using Workflow message in DSO classes.

mwoodiupui commented on dspace-api/src/main/java/org/dspace/xmlworkflow/XmlWorkflowServiceImpl.java:
```diff
@@ -631,6 +638,58 @@ protected void notifyOfArchive(Context context, Item item, Collection coll)
+        if (StringUtils.isNotBlank(title)) {
+            return title;
+        } else {
+            return I18nUtil.getMessage("org.dspace.workflow.WorkflowManager.untitled ");
```
This was also used in CommunityServiceImpl and CollectionServiceImpl, so I split it into two messages with more sensible names.

mwoodiupui commented on dspace-api/src/main/resources/spring/spring-dspace-addon-workflow-services.xml:
```diff
@@ -0,0 +1,86 @@
+        <constructor-arg type='java.lang.String'
+                         value='noaction'/>
+        <property name='processingAction'>
+            <bean class='org.dspace.xmlworkflow.state.actions.processingaction.NoAction'/>
```
The document has been moved into `test/resources`.

tdonohue commented on dspace-api/src/test/java/org/dspace/curate/XmlWorkflowCuratorServiceImplTest.java:
```diff
@@ -0,0 +1,277 @@
+        dspaceConfiguration.addPropertyValue(CURATION_TASK_MAP_NAME,
+                NoOpCurationTask.class.getCanonicalName() + " = " + TASK_NOOP);
+
+        // XXX Try to make Builders work in unit test.
+        AbstractBuilder.init();
```
@mwoodiupui : Just to second this thought (copying what I mentioned in Slack)... as has been discussed in #3055, we are trying to move away from Unit Tests creating any content.  So, the fact that this test class extends `AbstractUnitTests` _and uses Builders_ is a bit of an anti-pattern.  (It's something we are trying to move away from.)

So, there are a few options here... either we look at refactoring this test to no longer rely on creating content (if that's even possible), or we could look at simply changing this to an IT and extending `AbstractIntegrationTestWithDatabase` (which is where Builders are welcome).  Or, we could just keep this as is for now & clean it up as part of #3055 if neither of those options are easy right now.

mwoodiupui replied:
A better test would indeed be a welcome development.  You're seeing some of the rubble left after two weeks of bafflement, frustration and outrage as design after design failed, telling me that objects I had just created did not exist.  The current test code is in the "at last, it works! don't touch anything!" state.  The interior of the configurable workflow system is deeply mysterious, at least to me, and I hadn't found any enlightening examples of working with it.  I suppose this means that I should go tramping through the entire workflow subsystem, code machete in hand, and write a lot of documentation when the design becomes clear to me.

I'm happy to hear that there is movement to better organize our test practices.  This class was changed from IT to UT late in the game, when I felt that the IT framework was fighting me and I needed to simplify.  I stuck with the Builders because they presumably know the proper sequence of actions to set up the necessary objects, and it was becoming clear that I was missing some of that.  Possibly for UTs we need a parallel set of Maker classes to build in-memory objects *not* backed by the database, if we are going to continue preventing arbitrary code from instantiating model objects.  We also should do some refactoring to make some tested classes more testable.

tdonohue replied:
@mwoodiupui : Would you be willing to try changing this to extend `AbstractIntegrationTestWithDatabase` one last time?  I'm very curious why this would _work_ as an `AbstractUnitTest` with Builders, but _not work_ as an `AbstractIntegrationTestWithDatabase` with Builders.  That seems unexpected.

mwoodiupui replied:
I've scrapped that whole test suite and written a new IT class, extending `AbstractIntegrationTestWith Database`.  I was helped by having an example to look at.

Unfortunately I forgot why I was writing the test in the REST layer, and when the resulting code didn't seem to have any necessary connection with REST I factored it out and moved it back to dspace-api.

In the course of refactoring I discovered that (a) I needed two Spring initializers, which I copied from `dspace-server-webapp` to `dspace-api`, and (b) those initializers, while they say "Spring Boot" all over them, really don't depend on Boot at all.  I have wondered whether they really ought to be part of the scaffolding in `dspace-services`.  At any rate there's no reason why they can't be available to our whole constellation of projects, and having JUnit run them makes the test code quite a bit cleaner than trying to inject dependencies manually.

mwoodiupui committed:
[DS-2058] Merge curation test workflow into main workflow test configuration.

mwoodiupui committed:
[DS-2058] More documentation, small cleanups.

mwoodiupui committed:
[DS-2058] Less restrictive, more nearly correct Handle regex for workflow configuration.

mwoodiupui committed:
[DS-2058] Create a namespace for workflow-curation so my IDE can validate curation configurations.

mwoodiupui committed:
[DS-2058] Write, improve documentation of workflow classes in hope of understanding them better.

I suspect that in org.dspace.xmlworkflow.state.Step the word "valid" means
"conditions for transition to a different step are satisfied" or, more
concisely, "completed".

mwoodiupui committed:
[DS-2058] Attempt to curate workflow tasks in *all* the right places.

Workflow curation should happen at entry to a step, *before* looking for
manual tasks to send to reviewers.

Previous refactoring has split the old code into two pieces, so duplicate
the curation code to cover entry to the workflow as well as advancement
between steps.

mwoodiupui committed:
[DS-2058] WIP:  new test for workflow curation.

mwoodiupui committed:
[DS-2058] Extend and improve documentation.

mwoodiupui committed:
[DS-2058] MarkerTask should be in a curation tasks directory.

mwoodiupui committed:
[DS-2058] Move workflow curation testing to dspace-api since it has no essential connection with REST.

This patch duplicates the Spring/JUnit initializer classes from
dspace-server-webapp.  We should share a single set.  Possibly they should
live in dspace-services where the fundamental Spring support is found.

mwoodiupui committed:
[DS-2058] Finish removing workflow curation test from dspace-server-webapp.

This should completely revert the changes made to this artifact for workflow
curation.

mwoodiupui committed:
[DS-2058] Curator should not NPE if someone forgot to set a Reporter.

mwoodiupui committed:
[DS-2058] Fix misbehavior that was found in testing.

mwoodiupui committed:
[DS-2058] workflow-curation has a declared namespace now.

mwoodiupui committed:
[DS-2058] Avoid index out of bounds when no tasks match.

Also remove incorrect unit test.

mwoodiupui committed:
[DS-2058] Remove a missed added attribute, once again unused.

Some harmless cleanups remain, contrary to previous commit's comment.

mwoodiupui committed:
[DS-2058] Remove now-unused injected attribute.

mwoodiupui committed:
[DS-2058] Document the CollectionBuilder.

mwoodiupui committed:
[DS-2058] Test with the first non-default workflow as stated, not the last.

mwoodiupui committed:
[DS-2058] Tidy up for Checkstyle.

tdonohue reviewed (approved):
@mwoodiupui :  This looks great to me now.  I appreciate all the hard work you put into better code comments / documentation. The new ITs look good.  

I tested this by enabling the `noop` task during the `reviewstep` (which is the first step by default).  After submitting a new item into workflow, I saw this in my logs showing that the [`NoOpCurationTask`](https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/ctask/general/NoOpCurationTask.java) ran successfully
```
2021-06-22 19:25:46,375 INFO  org.dspace.curate.Curator @ Curation task: noop performed on: workflow item: ea39736e-2379-468f-b1d1-0b43cab1efb8 with status: 0. Result: 'No operation performed on null'
2021-06-22 19:25:46,377 INFO  org.dspace.curate.XmlWorkflowCuratorServiceImpl @ Curation tasks over item ea39736e-2379-468f-b1d1-0b43cab1efb8 for step reviewstep report:%nNo operation performed on null
```
(As a sidenote, in those logs, we can see some curation tasks report their result based on Item Handle, which comes back as `null`.  But, that's not a fault of this PR...it's a flaw in those Curation Tasks themselves)

All in all, I'm in favor of moving this forward, as it's working and is a major improvement.  That said, I have a few very minor suggestions to improve code comments inline below...

abollini commented on dspace-api/src/main/java/org/dspace/xmlworkflow/service/XmlWorkflowService.java:
```diff
@@ -41,39 +41,171 @@
     public void deleteAllPooledTasks(Context c, XmlWorkflowItem wi) throws SQLException, AuthorizeException;
 
+    /**
+     * Deletes an EPerson from the task pool of a step.
```
probably the confusion is due to the fact that tasks are usually in 1:1 relation with eperson. That said I agree with @tdonohue it would be better to rephrase

mwoodiupui replied:
Well caught.  I don't know what I was thinking when I wrote that this was about `EPerson`s.  Probably I had been staring for too long at too much undocumented code.

mwoodiupui commented on dspace-api/src/main/java/org/dspace/xmlworkflow/service/XmlWorkflowService.java:
```diff
@@ -41,39 +41,171 @@
+     * Deletes a completed task of a step.
+     * @param c    current DSpace session.
+     * @param wi   the workflow item from which to remove the EPerson.
+     * @param task the task from which to remove the EPerson.
```
Agreed, will be fixed.

mwoodiupui commented on dspace-api/src/main/java/org/dspace/xmlworkflow/service/XmlWorkflowService.java:
```diff
@@ -41,39 +41,171 @@
+     * Deletes an EPerson from the task pool of a step.
+     * @param context current DSpace session.
+     * @param wi      the workflow item from which to remove the EPerson.
+     * @param task    the task from which to remove the EPerson.
```
Agreed, will be fixed.

mwoodiupui commented on dspace-api/src/main/java/org/dspace/xmlworkflow/XmlWorkflowServiceImpl.java:
```diff
@@ -443,11 +483,10 @@ public WorkflowActionConfig processOutcome(Context c, EPerson user, Workflow wor
                     | WorkflowException | WorkflowConfigurationException e) {
                 log.error("error while processing workflow outcome", e);
-                e.printStackTrace();
+                e.printStackTrace(); // XXX Why is this here?
```
Thank you both.  Removed.

mwoodiupui committed:
[DS-2058] Improve documentation as requested.

mwoodiupui committed:
[DS-2058] A little more documentation touch-up.

tdonohue commented:
As discussed in today's meeting, merging this as-is.  I've tested and am satisfied... @abollini gave this at least a quick look (based on his feedback above).  Others on the DSpace 7 team were fine with merging immediately. Thanks @mwoodiupui again for your hard work here!
